src/main/java/org/mule/extension/sftp/internal/source/SftpDirectoryListener.java

```diff
@@ -0,0 +1,461 @@
+import org.mule.runtime.api.store.ObjectStoreException;
+import org.mule.runtime.api.store.ObjectStoreManager;
+import org.mule.runtime.api.store.ObjectStoreSettings;
+import org.mule.runtime.extension.api.annotation.Alias;
+import org.mule.runtime.extension.api.annotation.execution.OnError;
+import org.mule.runtime.extension.api.annotation.execution.OnSuccess;
+import org.mule.runtime.extension.api.annotation.execution.OnTerminate;
+import org.mule.runtime.extension.api.annotation.param.Config;
+import org.mule.runtime.extension.api.annotation.param.Connection;
+import org.mule.runtime.extension.api.annotation.param.MediaType;
+import org.mule.runtime.extension.api.annotation.param.Optional;
+import org.mule.runtime.extension.api.annotation.param.Parameter;
+import org.mule.runtime.extension.api.annotation.param.ParameterGroup;
+import org.mule.runtime.extension.api.annotation.param.display.DisplayName;
+import org.mule.runtime.extension.api.annotation.param.display.Summary;
+import org.mule.runtime.extension.api.annotation.source.BackPressure;
+import org.mule.runtime.extension.api.runtime.operation.Result;
+import org.mule.runtime.extension.api.runtime.source.Source;
+import org.mule.runtime.extension.api.runtime.source.SourceCallback;
+import org.mule.runtime.extension.api.runtime.source.SourceCallbackContext;
+
+import java.io.InputStream;
+import java.nio.file.Path;
+import java.time.LocalDateTime;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.Lock;
+import java.util.function.Predicate;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Polls a directory looking for files that have been created on it. One message will be generated for each file that is found.
+ * <p>
+ * The key part of this functionality is how to determine that a file is actually new. There're three strategies for that:
+ * <ul>
+ * <li>Set the <i>autoDelete</i> parameter to <i>true</i>: This will delete each processed file after it has been processed,
+ * causing all files obtained in the next poll to be necessarily new</li>
+ * <li>Set <i>moveToDirectory</i> parameter: This will move each processed file to a different directory after it has been
+ * processed, achieving the same effect as <i>autoDelete</i> but without loosing the file</li>
+ * <li></li>
+ * <li>Use the <i>watermarkMode</i> parameter to only pick files that have been created/updated after the last poll was
+ * executed.</li>
+ * </ul>
+ * <p>
+ * A matcher can also be used for additional filtering of files.
+ *
+ * @since 1.1
+ */
+@MediaType(value = ANY, strict = false)
+@DisplayName("On New File")
+@Summary("Triggers when a new file is created in a directory")
+@Alias("listener")
+@BackPressure(defaultMode = WAIT, supportedModes = WAIT)
+public class SftpDirectoryListener extends Source<InputStream, SftpFileAttributes> {
```

Comment:
no chance of refactoring to a something in file-commons that can be extended by ftp and sftp?

Reply:
i didn't do that on purpose. Can explain why off line